File: clients/web/src/app/services/shell-nav.util.ts
import type { MenuItem } from '../api/emcap-client';

export interface ModuleNavGroup {
  moduleCode: string;
  moduleLabel: string;
  items: MenuItem[];
}

export interface PlatformNavLink {
  label: string;
  labelKey?: string;
  route: string;
  visible: boolean;
}

const PLATFORM_MODULE_KEYS = new Set(['workflow', 'payments', 'notifications', 'ai']);

export function hasPermission(permissions: string[], required: string): boolean {
  if (permissions.includes('*.*') || permissions.includes(required)) {
    return true;
  }
  for (const entry of permissions) {
    if (entry.endsWith('.*')) {
      const prefix = entry.slice(0, -1);
      if (required.startsWith(prefix)) {
        return true;
      }
    }
  }
  return false;
}

export function menuPermission(menu: MenuItem): string {
  if (menu.permission && menu.permission !== 'read') {
    return menu.permission;
  }
  return `${menu.entity_code.toLowerCase()}.read`;
}

export function isModuleEnabled(
  moduleCode: string,
  modules: Record<string, { enabled?: boolean }> | undefined,
): boolean {
  const key = moduleCode.toLowerCase();
  if (!PLATFORM_MODULE_KEYS.has(key)) {
    return true;
  }
  return modules?.[key]?.enabled !== false;
}

export function filterMenus(
  menus: MenuItem[],
  userPermissions: string[],
  modules: Record<string, { enabled?: boolean }> | undefined,
): MenuItem[] {
  return menus.filter((menu) => {
    if (!isModuleEnabled(menu.module, modules)) {
      return false;
    }
    return hasPermission(userPermissions, menuPermission(menu));
  });
}

export function groupMenusByModule(menus: MenuItem[]): ModuleNavGroup[] {
  const map = new Map<string, MenuItem[]>();
  for (const menu of menus) {
    const code = menu.module || 'platform';
    const list = map.get(code) ?? [];
    list.push(menu);
    map.set(code, list);
  }
  return Array.from(map.entries())
    .sort(([left], [right]) => left.localeCompare(right))
    .map(([moduleCode, items]) => ({
      moduleCode,
      moduleLabel: formatModuleLabel(moduleCode),
      items: [...items].sort((a, b) => a.label.localeCompare(b.label)),
    }));
}

export function formatModuleLabel(code: string): string {
  if (code === 'platform') {
    return 'Platform';
  }
  return code
    .toLowerCase()
    .split('_')
    .map((part) => part.charAt(0).toUpperCase() + part.slice(1))
    .join(' ');
}

export function buildPlatformLinks(
  modules: Record<string, { enabled?: boolean }> | undefined,
  userPermissions: string[] = ['*.*'],
): PlatformNavLink[] {
  const links: PlatformNavLink[] = [
    {
      labelKey: 'platform.workflow.title',
      label: 'Workflow tasks',
      route: '/app/workflow',
      visible: modules?.workflow?.enabled !== false,
    },
    { labelKey: 'platform.reports.title', label: 'Reports', route: '/app/reports', visible: true },
    {
      labelKey: 'platform.dashboards.title',
      label: 'Dashboards',
      route: '/app/dashboards',
      visible: true,
    },
    {
      labelKey: 'platform.notifications.title',
      label: 'Notifications',
      route: '/app/notifications',
      visible: modules?.notifications?.enabled !== false,
    },
    { labelKey: 'platform.account.title', label: 'Account', route: '/app/account', visible: true },
    {
      labelKey: 'platform.assistant.title',
      label: 'Assistant',
      route: '/app/assistant',
      visible: modules?.ai?.enabled === true,
    },
  ];
  if (
    hasPermission(userPermissions, 'admin.users.read') ||
    hasPermission(userPermissions, 'admin.roles.read') ||
    hasPermission(userPermissions, 'admin.*')
  ) {
    links.push({ labelKey: 'nav.admin', label: 'Admin', route: '/app/admin/users', visible: true });
  }
  if (hasPermission(userPermissions, 'admin.settings.read') || hasPermission(userPermissions, 'admin.*')) {
    links.push({ labelKey: 'nav.settings', label: 'Settings', route: '/app/settings', visible: true });
  }
  return links;
}

File: clients/mobile/lib/utils/shell_nav_util.dart
class MenuItem {
  MenuItem({
    required this.code,
    required this.label,
    required this.entityCode,
    required this.module,
    this.icon,
    this.permission,
  });

  final String code;
  final String label;
  final String entityCode;
  final String module;
  final String? icon;
  final String? permission;

  factory MenuItem.fromJson(Map<String, dynamic> json) {
    return MenuItem(
      code: '${json['code']}',
      label: '${json['label']}',
      entityCode: '${json['entity_code']}',
      module: '${json['module'] ?? 'platform'}',
      icon: json['icon'] as String?,
      permission: json['permission'] as String?,
    );
  }
}

class ModuleNavGroup {
  ModuleNavGroup({required this.moduleCode, required this.moduleLabel, required this.items});

  final String moduleCode;
  final String moduleLabel;
  final List<MenuItem> items;
}

class PlatformNavLink {
  PlatformNavLink({required this.key, required this.label, required this.visible});

  final String key;
  final String label;
  final bool visible;
}

const _platformModuleKeys = {'workflow', 'payments', 'notifications', 'ai'};

bool hasPermission(List<String> permissions, String required) {
  if (permissions.contains('*.*') || permissions.contains(required)) {
    return true;
  }
  for (final entry in permissions) {
    if (entry.endsWith('.*')) {
      final prefix = entry.substring(0, entry.length - 1);
      if (required.startsWith(prefix)) {
        return true;
      }
    }
  }
  return false;
}

String menuPermission(MenuItem menu) {
  if (menu.permission != null && menu.permission != 'read') {
    return menu.permission!;
  }
  return '${menu.entityCode.toLowerCase()}.read';
}

bool isModuleEnabled(String moduleCode, Map<String, dynamic>? modules) {
  final key = moduleCode.toLowerCase();
  if (!_platformModuleKeys.contains(key)) {
    return true;
  }
  final toggle = modules?[key];
  if (toggle is Map) {
    return toggle['enabled'] != false;
  }
  return true;
}

List<MenuItem> filterMenus(
  List<MenuItem> menus,
  List<String> userPermissions,
  Map<String, dynamic>? modules,
) {
  return menus.where((menu) {
    if (!isModuleEnabled(menu.module, modules)) {
      return false;
    }
    return hasPermission(userPermissions, menuPermission(menu));
  }).toList();
}

List<ModuleNavGroup> groupMenusByModule(List<MenuItem> menus) {
  final map = <String, List<MenuItem>>{};
  for (final menu in menus) {
    map.putIfAbsent(menu.module, () => []).add(menu);
  }
  final entries = map.entries.toList()..sort((a, b) => a.key.compareTo(b.key));
  return entries
      .map(
        (entry) => ModuleNavGroup(
          moduleCode: entry.key,
          moduleLabel: formatModuleLabel(entry.key),
          items: [...entry.value]..sort((a, b) => a.label.compareTo(b.label)),
        ),
      )
      .toList();
}

String formatModuleLabel(String code) {
  if (code == 'platform') {
    return 'Platform';
  }
  return code
      .toLowerCase()
      .split('_')
      .map((part) => part.isEmpty ? part : '${part[0].toUpperCase()}${part.substring(1)}')
      .join(' ');
}

List<PlatformNavLink> buildPlatformLinks(
  Map<String, dynamic>? modules,
  List<String> userPermissions,
) {
  final links = <PlatformNavLink>[
    PlatformNavLink(
      key: 'workflow',
      label: 'Workflow tasks',
      visible: (modules?['workflow'] as Map?)?['enabled'] != false,
    ),
    PlatformNavLink(key: 'reports', label: 'Reports', visible: true),
    PlatformNavLink(key: 'dashboards', label: 'Dashboards', visible: true),
    PlatformNavLink(
      key: 'notifications',
      label: 'Notifications',
      visible: (modules?['notifications'] as Map?)?['enabled'] != false,
    ),
    PlatformNavLink(key: 'account', label: 'Account', visible: true),
    PlatformNavLink(
      key: 'assistant',
      label: 'Assistant',
      visible: (modules?['ai'] as Map?)?['enabled'] == true,
    ),
  ];
  if (hasPermission(userPermissions, 'admin.users.read') ||
      hasPermission(userPermissions, 'admin.roles.read') ||
      hasPermission(userPermissions, 'admin.*')) {
    links.add(PlatformNavLink(key: 'admin-users', label: 'Admin users', visible: true));
    links.add(PlatformNavLink(key: 'admin-roles', label: 'Admin roles', visible: true));
    links.add(PlatformNavLink(key: 'admin-permissions', label: 'Permissions', visible: true));
  }
  if (hasPermission(userPermissions, 'admin.security.read') || hasPermission(userPermissions, 'admin.*')) {
    links.add(PlatformNavLink(key: 'admin-security', label: 'Security policies', visible: true));
  }
  if (hasPermission(userPermissions, 'admin.settings.read') || hasPermission(userPermissions, 'admin.*')) {
    links.add(PlatformNavLink(key: 'settings', label: 'Settings', visible: true));
  }
  return links.where((link) => link.visible).toList();
}

List<String> extractUserPermissions(Map<String, dynamic> me) {
  final perms = me['permissions'];
  if (perms is List) {
    return perms.map((item) => '$item').toList();
  }
  return ['*.*'];
}

Map<String, dynamic>? extractModuleToggles(Map<String, dynamic> config) {
  final modules = config['modules'];
  if (modules is Map) {
    return Map<String, dynamic>.from(modules);
  }
  return null;
}

/// API returns `tenants` as a map keyed by tenant id; tests may use a list.
List<Map<String, dynamic>> parseTenantEntries(Map<String, dynamic> payload) {
  final raw = payload['tenants'];
  if (raw is List) {
    return [
      for (final item in raw)
        if (item is Map) Map<String, dynamic>.from(item),
    ];
  }
  if (raw is Map) {
    return [
      for (final entry in raw.entries)
        {
          'id': entry.key,
          if (entry.value is Map) ...Map<String, dynamic>.from(entry.value as Map),
        },
    ];
  }
  return [];
}

class ShellNavEntryRef {
  const ShellNavEntryRef({
    required this.key,
    required this.label,
    this.groupLabel,
  });

  final String key;
  final String label;
  final String? groupLabel;
}

class RailNavSlot {
  const RailNavSlot.header(this.headerLabel) : key = null, label = null;

  const RailNavSlot.destination({required this.key, required this.label}) : headerLabel = null;

  final String? headerLabel;
  final String? key;
  final String? label;

  bool get isHeader => headerLabel != null;
}

List<RailNavSlot> buildRailNavSlots(List<ShellNavEntryRef> entries) {
  final slots = <RailNavSlot>[];
  String? currentGroup;
  for (final entry in entries) {
    if (entry.groupLabel != null && entry.groupLabel != currentGroup) {
      currentGroup = entry.groupLabel;
      slots.add(RailNavSlot.header(currentGroup!));
    }
    slots.add(RailNavSlot.destination(key: entry.key, label: entry.label));
  }
  return slots;
}

List<String> moduleHeaderLabels(List<RailNavSlot> slots) {
  return [
    for (final slot in slots)
      if (slot.isHeader) slot.headerLabel!,
  ];
}
